src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -253,27 +254,96 @@ ExpressionEvaluator::_evaluate_in_expression<ExpressionEvaluator::Bool>(const In
-      if ((element->data_type() == DataType::String) == left_is_string) {
-        type_compatible_elements.emplace_back(element);
+    bool all_elements_are_values_of_left_type = true;
+    resolve_data_type(left_expression.data_type(), [&](const auto left_data_type_t) {
+      using LeftDataType = typename decltype(left_data_type_t)::type;
+
+      for (const auto& element : array_expression.elements()) {
+        if ((element->data_type() == DataType::String) == left_is_string) {
+          type_compatible_elements.emplace_back(element);
+        }
+
+        if (element->type != ExpressionType::Value) {
+          all_elements_are_values_of_left_type = false;
+        } else {
+          const auto& value_expression = std::static_pointer_cast<ValueExpression>(element);
+          if (value_expression->value.type() != typeid(LeftDataType)) all_elements_are_values_of_left_type = false;
+        }
       }
-    }
+    });
 
     if (type_compatible_elements.empty()) {
-      // `5 IN ()` is FALSE as is `NULL IN ()`
+      // `5 IN ()` is FALSE
       return std::make_shared<ExpressionResult<ExpressionEvaluator::Bool>>(std::vector<ExpressionEvaluator::Bool>{0});
     }
 
+    // If all elements of the list are simple values (e.g., `IN (1, 2, 3)`), iterate over the column and directly
+    // compare the left value with the values in the list. A binary search is used because of its algorithmic beauty
+    // (and for reeeeally long lists, as they might come from ORMs).
+    //
+    // If we can't store the values in a vector (because they are too complex), we translate the IN clause to a series
+    // of ORs:
+    // "a IN (x, y, z)"   ---->   "a = x OR a = y OR a = z"
+    // The first path is faster, while the second one is more flexible.
+    if (all_elements_are_values_of_left_type) {
+      _resolve_to_expression_result_view(left_expression, [&](const auto& left_view) {
+        using LeftDataType = typename std::decay_t<decltype(left_view)>::Type;
+
+        // Above, we have ruled out NULL on the left side, but the compiler does not know this yet
+        if constexpr (!std::is_same_v<LeftDataType, NullValue>) {
+          std::vector<LeftDataType> right_values;
+          right_values.reserve(type_compatible_elements.size());
+          for (const auto& expression : type_compatible_elements) {
+            const auto& value_expression = std::static_pointer_cast<ValueExpression>(expression);
+            right_values.emplace_back(boost::get<LeftDataType>(value_expression->value));
+          }
+          std::sort(right_values.begin(), right_values.end());
+
+          result_values.resize(left_view.size());
+          if (left_expression.is_nullable()) {
+            result_nulls.resize(left_view.size());
+          }
+
+          for (auto chunk_offset = ChunkOffset{0}; chunk_offset < left_view.size(); ++chunk_offset) {
+            if (left_expression.is_nullable() && left_view.is_null(chunk_offset)) {
```

Comment:
Why do `left_expression.is_nullable()` which is potentially a non-trivial lookup (think `a + b+ c IN (...)`) instead of `left_view.is_nullable()` which is either a trivial runtime lookup (`ExpressionResultLiteral`) or known at compile time (`ExpressionResultNonNullSeries` and `ExpressionResultNullableSeries`)

Reply:
Because I didn't know better. Any chance of communicating the difference more explicitly?

Reply:
"communicating the difference more explicitly" in "a comment on this PR" or in general "as documentation in the code"?

Reply:
In general, making sure others don't make the same mistake.